Update index.html

KipOmaha opened the pull request:
removed unfinished sentence.
Added `Creating your own Extension` Heading did not finish content.

KipOmaha commented:
closing pull request https://github.com/x-tag/x-tag.github.io/pull/16
this PR contains all the changes from the #16

KipOmaha commented on index.html:
```diff
@@ -1,725 +1,1449 @@
-    // the disabled="" attribute. Because it is declared as a boolean, 
-    // all values passed into its getter/setter will be a Boolean.
-    this.firstElementChild.disabled = value;
+  get 'mode::attr'(){ 
```
So basically what I did here was use it the way your template attribute uses it.  I wasn't sure about this myself, did you want to include any  content on these types of uses?  

In the future I can bring up issues before making a PR.

KipOmaha replied:
Minus the promise api bit of course.

KipOmaha replied:
I got that mistaken a bit...the render() api will always write the newest template that you summoned into existence...which I thought it wouldn't.  Maybe a line in the docs should be added just in case others misunderstand that templates get replaced rather than appended onto the targeted node.

I.E. the mode attribute itself was so that you could create different types and than define different modes for those types. I have a bit more for x-tag components concerning types and modes.  

I have to push an update to my dashboard build, when I do that I could create a branch with the build that utilizes the x-tags type and mode extension, if you want. It would be helpful, cause I could use some help concerning the event API being used in the extensions being defined. [i.e. they are not firing when added to an extension object]

KipOmaha replied:
I did more testing with the extensions and got the src plugin running, I guess something maybe out of place somewhere I overlooked.

KipOmaha replied:
Duh I just found it.

KipOmaha replied:
I'm sorry for bombing this thread but this is annoying me...I have bit of code working the way its supposed to but it matters what order I put the extensions in. For instance this works =>
`XTagElement.extensions("mode","typed",...)`

but not

`XTagElement.extensions("typed","mode",...)`

KipOmaha replied:
Is this a bug or is their a reason I'm not seeing?

KipOmaha commented on index.html:
```diff
@@ -604,18 +565,57 @@ <h4>Create your own Pseudo:</h4>
-<h4>To an Extended Voyage</h4>
+<pre class="language-javascript"><code>// Lets create our X-Tag extension.
+
+xtag.extensions.lifecycle = {
```
I agree, I threw it in there from my own v2 extensions that I'm building.  How about making a  `xtag.extensions.jenny = { }' example?

KipOmaha replied:
lol sorry `xtag.extensions.foo`

KipOmaha replied:
I'm going to commit a different example today.

csuwildcat commented:
I agree that the test examples are probably a good idea for this.

On Wed, Apr 25, 2018, 9:19 PM KIP <notifications@github.com> wrote:

> I think its a good idea to commit an extensions example from the jasmine
> testing what do you think? I think that the examples I used are good but in
> a hacky sort of way....
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/x-tag/x-tag.github.io/pull/19#issuecomment-384508273>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAICyuL1WajnEx1hGwoausd8VGwQ2oiUks5tsUrCgaJpZM4TLaYY>
> .
>

csuwildcat commented:
This could be a bug, I'm not sure why this would happen. Can you create a
simple Codepen out of this example and link it here?

On Thu, Apr 26, 2018, 7:00 PM KIP <notifications@github.com> wrote:

> *@KipOmaha* commented on this pull request.
> ------------------------------
>
> In index.html
> <https://github.com/x-tag/x-tag.github.io/pull/19#discussion_r184576173>:
>
> >
> -xtag.create('x-bar', class extends XTagElement {
> -  '::template(true)' (){ return '&lt;input /&gt;' }
> -  set 'disabled::attr(boolean)' (value){
> -    // The ::attr extended property links node.disabled to gets/sets of
> -    // the disabled="" attribute. Because it is declared as a boolean,
> -    // all values passed into its getter/setter will be a Boolean.
> -    this.firstElementChild.disabled = value;
> +  get 'mode::attr'(){
>
> Is this a bug or is their a reason I'm not seeing?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/x-tag/x-tag.github.io/pull/19#discussion_r184576173>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAICymlK7aCwUo3jBxhf22mfDi7alqPrks5tsnutgaJpZM4TLaYY>
> .
>

KipOmaha commented:
Of course.! :)

KipOmaha commented:
https://codepen.io/kipomaha/pen/xjRKwa

csuwildcat commented:
The test case is pretty large, and I can't really track the error well,
given there are many, many JS includes and it is hard to understand the
loading sequence. Can you do something that shows where the x-tag lib is
being loaded, vs the Pen code that throws? You are loading the lib before
the custom code, correct?

On Thu, Apr 26, 2018 at 7:15 PM, KIP <notifications@github.com> wrote:

> https://codepen.io/kipomaha/pen/xjRKwa
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/x-tag/x-tag.github.io/pull/19#issuecomment-384843311>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAICykrYTloOqTTP-QcZbzH_XiT7ngWCks5tsn9VgaJpZM4TLaYY>
> .
>

csuwildcat commented:
It looks like the Pen is failing to load the base X-Tag JS, which is making
it impossible to debug any subsequent issue that may be in play further
down when you use the lib/extensions code.

On Fri, Apr 27, 2018 at 10:46 AM, Daniel Buchner <danieljb2@gmail.com>
wrote:

> The test case is pretty large, and I can't really track the error well,
> given there are many, many JS includes and it is hard to understand the
> loading sequence. Can you do something that shows where the x-tag lib is
> being loaded, vs the Pen code that throws? You are loading the lib before
> the custom code, correct?
>
> On Thu, Apr 26, 2018 at 7:15 PM, KIP <notifications@github.com> wrote:
>
>> https://codepen.io/kipomaha/pen/xjRKwa
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/x-tag/x-tag.github.io/pull/19#issuecomment-384843311>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAICykrYTloOqTTP-QcZbzH_XiT7ngWCks5tsn9VgaJpZM4TLaYY>
>> .
>>
>
>

KipOmaha commented:
I am new to codepen. I just started the account, so I dumped a bunch and started reorganizing.
The Problem was resolved, though. 
Sorry for the false alarm...long story short I thought I had fixed all the part in the code but missed an occurrence of it in the mode.js file. 
MORE DETAILS BELOW. It is a strange occurrence and is a stable way to build a mixin if you had a diff library that you wanted to incorporate.

When I started to put it onto code pen it became apparent to me what had happened inadvertently.

When I created the mixin for the type extension I went and did:

`mixin: (base) => class extends XTagElement {}`
instead of 
`mixin: (base) => class extends base {}`

Which was a problem I had earlier [I thought for sure I had fixed them all] but didn't recognize the symptoms, because the `oncompile` , `onparse`, and `onconstruct` were  working.

KipOmaha commented:
here is the code pen for the offending file. https://codepen.io/kipomaha/pen/gzLKea?editors=1010

KipOmaha commented:
I guess what I did was create a custom class mixin in away.

KipOmaha commented:
https://codepen.io/kipomaha/full/xjRKwa  <= example of my error....nobody should ever do this.......

KipOmaha commented:
@csuwildcat the example of my error is their if you wanted to see what it causes.

KipOmaha commented:
Open the console [may want to drag it so its taller] and you will see the log. `x-bili` doesn't see the `bili` attribute

csuwildcat commented:
@KipOmaha this is the output I see on that page:

```console
pen.js:28   hi onParse.
pen.js:8     hi onParse.
pen.js:36   testin attr bil.
pen.js:37   hi onConstruct
pen.js:16   testin attr bili.
pen.js:17   hi onConstruct
pen.js:28   hi onParse.
pen.js:8     hi onParse.
pen.js:36   testin attr bil.
pen.js:37   hi onConstruct
pen.js:16   undefined
pen.js:17   hi onConstruct
```

Is the issue line 16 where the value is `undefined`?

KipOmaha commented:
So the code runs the same except that the `XTagElement.extensions("bil","bili",...)` is switched up so that the custom extensions are presented as parameters in a different order in each custom element, @csu wildcat. 
The order causes it to be undefined in one custom element and not the other.  It's kind of confusing...sorry :(

KipOmaha commented:
@csuwildcat I made changes to the test case.  Below is the log.  Okay so the extension get executed in the order that they appear in the element definitions. So the parse, compile, and construct methods will get executed in the order they appear not in the extension paramets but in the object.  I still don't quite get why the order of this matters...I mean the mistake in the `mixin` I made should have thrown an error I would have thought intuitively but it runs in some cases making it hard to identify when boxing with multiple JavaScript element definition files. 
```
 Parsing `bib` extension.   ------   xjRKwa (85,5)

 Parsing `bar` extension.   ------   xjRKwa (66,5)

 Parsing `bal` extension.   ------   xjRKwa (104,5)

 Compiling `bib` extension.   ------   xjRKwa (88,5)

 Compiling `bar` extension   ------   xjRKwa (69,5)

 Compiling `bal` extension.   ------   xjRKwa (107,5)

 Constructing `bib` extension for:    ------   xjRKwa (91,5)

<x-foo bal="x-foo" bib="x-foo" bili="x-foo"></x-foo>

 Testing attr getter `bil`.   ------   xjRKwa (93,5)

 Result:true   ------   xjRKwa (94,16)

 Constructing `bar` extension for:    ------   xjRKwa (72,5)


<x-foo bal="x-foo" bib="x-foo" bili="x-foo"></x-foo>

 Testing attr getter bili.   ------   xjRKwa (74,5)

 Result:true   ------   xjRKwa (75,17)

 Constructing `bal` extension for:    ------   xjRKwa (110,5)


<x-foo bal="x-foo" bib="x-foo" bili="x-foo"></x-foo>

 Testing attr bal.   ------   xjRKwa (112,5)

 Result:true   ------   xjRKwa (113,16)

 Parsing `bib` extension.   ------   xjRKwa (85,5)

 Parsing `bar` extension...   ------   xjRKwa (66,5)

 Parsing `bal` extension.   ------   xjRKwa (104,5)

Compiling `bib` extension.  ------   xjRKwa (88,5)

 Compiling `bar` extension  ------   xjRKwa (69,5)

 Compiling `bal` extension.  ------   xjRKwa (107,5)

 Constructing `bib` extension for:   ------   xjRKwa (91,5)

<x-tup bal="x-tup" bib="x-tup" bili="x-tup"></x-tup>

 Testing attr getter `bil`.  ------   xjRKwa (93,5)

 Result:true  ------   xjRKwa (94,16)

 Constructing `bar` extension for:  ------    xjRKwa (72,5)

<x-tup bal="x-tup" bib="x-tup" bili="x-tup"></x-tup>

 undefined  ------   xjRKwa (74,5)

 Result:false  ------   xjRKwa (75,49)

 Constructing `bal` extension for:   ------   xjRKwa (110,5)
 
<x-tup bal="x-tup" bib="x-tup" bili="x-tup"></x-tup>

 Testing attr bal.  ------   xjRKwa (112,5)

 Result:true   ------   xjRKwa (113,16)
```
